pallets/gear-program/src/migrations.rs

```diff
@@ -0,0 +1,342 @@
+        assert_eq!(
+            StorageVersion::get::<Pallet<T>>(),
+            2,
+            "Can only upgrade from version 2"
+        );
+
+        let count = v2::ProgramStorage::<T>::iter().fold(0u64, |count, (program_id, program)| {
+            match program {
+                v2::Program::Terminated(_) | v2::Program::Exited(_) => {
+                    assert!(v2::MemoryPageStorage::<T>::iter_key_prefix(program_id)
+                        .next()
+                        .is_none());
+                }
+                v2::Program::Active(_) => (),
+            }
+
+            count + 1
+        });
+
+        Ok(count.encode())
+    }
+
+    fn on_runtime_upgrade() -> Weight {
+        let current = Pallet::<T>::current_storage_version();
+        let onchain = Pallet::<T>::on_chain_storage_version();
+
+        log::info!(
+            "🚚 Running migration with current storage version {current:?} / onchain {onchain:?}"
+        );
+
+        // 1 read for on chain storage version
+        let mut weight = T::DbWeight::get().reads(1);
+
+        if current == 3 && onchain == 2 {
+            ProgramStorage::<T>::translate(
+                |program_id, program: v2::Program<BlockNumberFor<T>>| {
+                    weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+
+                    Some(match program {
+                        v2::Program::Active(p) => {
+                            for (page, data) in v2::MemoryPageStorage::<T>::drain_prefix(program_id)
+                            {
+                                weight =
+                                    weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+
+                                MemoryPages::<T>::insert((program_id, MEMORY_INFIX, page), data);
+                            }
+
+                            Program::Active(common::ActiveProgram {
+                                allocations: p.allocations,
+                                pages_with_data: p.pages_with_data,
+                                gas_reservation_map: p.gas_reservation_map,
+                                code_hash: p.code_hash,
+                                code_exports: p.code_exports,
+                                static_pages: p.static_pages,
+                                state: p.state,
+                                expiration_block: p.expiration_block,
+                                memory_infix: MEMORY_INFIX,
+                            })
+                        }
+                        v2::Program::Exited(id) => Program::Exited(id),
+                        v2::Program::Terminated(id) => Program::Terminated(id),
+                    })
+                },
+            );
+
+            let _ = ResumeSessions::<T>::clear(u32::MAX, None);
+            let _ = v2::SessionMemoryPages::<T>::clear(u32::MAX, None);
+            let _ = PausedProgramStorage::<T>::clear(u32::MAX, None);
```

Comment:
And why should we clear paused program storage?

Reply:
same question about resume sessions

Reply:
Actually this is unnecessary since the rent logic is disabled at the moment. Will change that to assert+log::error